cmd/reflect/reflect.go

```diff
@@ -0,0 +1,160 @@
+// Copyright 2020 Authors of Hubble
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package reflect
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"os"
+
+	"github.com/cilium/hubble/pkg/defaults"
+	"github.com/golang/protobuf/proto"
+	"github.com/golang/protobuf/protoc-gen-go/descriptor"
+	"github.com/spf13/cobra"
+	"google.golang.org/grpc"
+	rpb "google.golang.org/grpc/reflection/grpc_reflection_v1alpha"
+)
+
+var (
+	serverURL string
+)
+
+// New returns the reflect command.
+func New() *cobra.Command {
+	reflectCmd := &cobra.Command{
+		Use:   "reflect",
+		Short: "Let Hubble reflect upon itself",
+		RunE: func(cmd *cobra.Command, args []string) error {
+			return runReflect()
+		},
+		Hidden: true,
+	}
+	reflectCmd.Flags().StringVarP(&serverURL, "server", "", defaults.GetDefaultSocketPath(), "URL to connect to server")
+	return reflectCmd
+}
+
+func runReflect() (err error) {
+	conn, err := grpc.Dial(serverURL, grpc.WithInsecure())
+	if err != nil {
+		return err
+	}
+	defer func() {
+		if closeErr := conn.Close(); closeErr != nil && err == nil {
+			err = closeErr
+		}
+	}()
+
+	ctx, cancel := context.WithTimeout(context.Background(), defaults.DefaultRequestTimeout)
+	defer cancel()
+
+	client, err := rpb.NewServerReflectionClient(conn).ServerReflectionInfo(ctx)
+	if err != nil {
+		return err
+	}
+	req := rpb.ServerReflectionRequest{
+		Host:           serverURL,
+		MessageRequest: &rpb.ServerReflectionRequest_ListServices{},
+	}
+	if err := client.Send(&req); err != nil {
+		return err
+	}
+	res, err := client.Recv()
+	if err != nil {
+		return err
+	}
+	services, ok := res.GetMessageResponse().(*rpb.ServerReflectionResponse_ListServicesResponse)
+	if !ok {
+		return fmt.Errorf("unexpected response: %v", res)
+	}
+	// same proto file can be imported multiple times from different places. keep track of proto filenames
+	// so that we don't print them multiple times.
+	visited := make(map[string]bool)
```

Comment:
`map[string]struct{}` is technically more efficient

Reply:
yeah i always forget about this 🤯